### PR TITLE
Review 7496: Latest Sketch Beta crashes when exporting selected Artboards to Zeplin

### PR DESCRIPTION
**Pull request by samdeane**
Fixes for 7496: Latest Sketch Beta crashes when exporting selected Artboards to Zeplin

Connect to BohemianCoding/Sketch#7496.

### CODE REVIEW

File: src/framework/mocha/MochaRuntime.m [hunk 1]
```diff
@@ -1131,12 +1131,18 @@ JSValueRef Mocha_getProperty(JSContextRef ctx, JSObjectRef object, JSStringRef p
 #pragma mark Mocha Objects
 
 static void MOObject_initialize(JSContextRef ctx, JSObjectRef jsObjectRepresentingBox) {
-    NSCAssert([((__bridge MOBox *)JSObjectGetPrivate(jsObjectRepresentingBox)) isKindOfClass:[MOBox class]], @"should have an associated box object");
+    NSDictionary* context = (__bridge NSDictionary *)(JSObjectGetPrivate(jsObjectRepresentingBox));
+    NSCAssert([context isKindOfClass:[NSDictionary class]], @"should have an dictionary with context");
+
+    MOBoxManager* manager = context[@"manager"];
+    id object = context[@"object"];
```
Comment (randomsequence) [inline src/framework/mocha/MochaRuntime.m]: I'd like to see these strings as constants or, even better, a custom object.

File: src/framework/mocha/Objects/MOBoxManager.m [hunk 2]
```diff
@@ -46,21 +55,26 @@ - (MOBox*)boxForObject:(id)object {
 - (JSObjectRef)makeBoxForObject:(id)object jsClass:(JSClassRef)jsClass {
     NSAssert([NSThread isMainThread], @"should be main thread");
     NSAssert(![object isKindOfClass:[MOBox class]], @"shouldn't box a box");
-    MOBox* box = [[MOBox alloc] initWithManager:self];
-    JSObjectRef jsObject = JSObjectMake(_context, jsClass, (__bridge void *)(box));
-    [box associateObject:object jsObject:jsObject];
+    NSDictionary* context = @{ @"manager" : self, @"object" : object };
```
Comment (randomsequence) [inline src/framework/mocha/Objects/MOBoxManager.m]: As above - constants, or a custom context object.

**Comment by randomsequence** [comment]
Reviewed with request.

File: src/framework/mocha/MochaRuntime.m (hunk 1 above)
Comment (samdeane) [inline src/framework/mocha/MochaRuntime.m]: Yeah, you're right - I was being lazy. Custom object it is...

**Comment by randomsequence** [comment]
Reviewed again. I don't really see what `-[MOBoxManagerBoxContext finishMakingBoxForObject:]` gains us? It seems to expose a method on `MOBoxManager` which `MOObject_initialize()` previously called directly.

**Comment by samdeane** [comment]
Better encapsulation, essentially. It allows MOBoxManagerBoxContext to remain completely opaque, and  allows the interaction between MOBoxManagerBoxContext and MOBoxManager to use a private API.

**Comment by samdeane** [comment]
I just wanted to try to avoid anyone thinking that it was ok to call `associateObject` on MOBoxManager directly. This way all you can do is call `finishMakingBoxForObject` on a context object, and the only way you can get one of those is via `makeBoxForObject` and the JS callback.

**Comment by randomsequence** [comment]
That makes sense. Good to QA :hand:

**Comment by pieteromvlee** [comment]
(Merged to develop)

**Pull request closed without merging**